src/Stratis.SmartContracts.Executor.Reflection/ContractInvocationResult.cs

```diff
@@ -15,32 +16,46 @@ private ContractInvocationResult(object result)
         }
 
-        private ContractInvocationResult(ContractInvocationErrorType errorType)
+        private ContractInvocationResult(ContractInvocationErrorType errorType, ContractErrorMessage errorMessage)
         {
             this.IsSuccess = false;
             this.InvocationErrorType = errorType;
-        }
-
-        private ContractInvocationResult(ContractInvocationErrorType errorType, Exception exception)
-        {
-            this.IsSuccess = false;
-            this.InvocationErrorType = errorType;
-            this.Exception = exception;
+            this.ErrorMessage = errorMessage;
         }
 
         public static ContractInvocationResult Success(object result)
         {
             return new ContractInvocationResult(result);
         }
 
+        /// <summary>
+        /// Return invocation failure for cases outside the execution of contract code.
+        /// </summary>
         public static ContractInvocationResult Failure(ContractInvocationErrorType errorType)
         {
-            return new ContractInvocationResult(errorType);
+            switch (errorType)
+            {
+                case ContractInvocationErrorType.MethodDoesNotExist:
+                    return new ContractInvocationResult(errorType,  new ContractErrorMessage("Method does not exist on contract."));
+                case ContractInvocationErrorType.MethodIsConstructor:
+                    return new ContractInvocationResult(errorType, new ContractErrorMessage("Attempted to invoke constructor on existing contract."));
+                case ContractInvocationErrorType.MethodIsPrivate:
+                    return new ContractInvocationResult(errorType, new ContractErrorMessage("Attempted to invoke private method."));
+                case ContractInvocationErrorType.ParameterCountIncorrect:
+                    return new ContractInvocationResult(errorType, new ContractErrorMessage("Incorrect number of parameters passed to method."));
+                case ContractInvocationErrorType.ParameterTypesDontMatch:
+                    return new ContractInvocationResult(errorType, new ContractErrorMessage("Parameters sent don't match expected method parameters."));
+                default:
+                    throw new NotSupportedException($"Should use either {nameof(Success)} or {nameof(ExecutionFailure)} for this ContractInvocationErrorType.");
+            }
         }
 
-        public static ContractInvocationResult Failure(ContractInvocationErrorType errorType, Exception exception)
+        /// <summary>
+        /// Return invocation failure for cases related to execution inside contract code.
+        /// </summary>
+        public static ContractInvocationResult ExecutionFailure(ContractInvocationErrorType errorType, Exception exception)
         {
-            return new ContractInvocationResult(errorType);
+            return new ContractInvocationResult(errorType, new ContractErrorMessage(exception.ToString()));
```

Comment:
Should we be using Exception.Message rather than Exception.ToString? ToString will be very verbose, including a stack trace.

Reply:
Yeah, as this is only for the non-consensus version of the receipts, for convenience we will store the whole thing.